### PR TITLE
configurable buffer sizes + custom headers

### PR DESCRIPTION
**Pull request by danielnorberg**

### CODE REVIEW

**Comment by danielnorberg** [comment]
@xerial PTAL

**Comment by danielnorberg** [comment]
@xerial friendly poke

File: src/main/java/com/treasuredata/client/AbstractTDClientBuilder.java [hunk 1]
```diff
@@ -62,6 +64,9 @@
     protected int connectTimeoutMillis = 15000;
     protected int idleTimeoutMillis = 60000;
     protected int connectionPoolSize = 64;
+    protected Multimap<String, String> headers = ImmutableMultimap.of();
+    protected Optional<Integer> requestBufferSize = Optional.absent();
+    protected Optional<Integer> responseBufferSize = Optional.absent();
```
Comment (xerial) [inline src/main/java/com/treasuredata/client/AbstractTDClientBuilder.java]: Just from curiosity; what is the ideal size for these buffers?

File: src/test/java/com/treasuredata/client/TestTDClientConfig.java [hunk 2]
```diff
@@ -199,4 +208,62 @@ public void canConvertToProperties()
...
+        assertThat(clientWithHeaders.httpClient.headers, is(equalTo(headers)));
+        assertThat(clientWithNoHeaders1.withHeaders(extraHeaders).httpClient.headers, is(equalTo(extraHeaders)));
+        assertThat(clientWithNoHeaders1.withHeaders(extraHeaders).withHeaders(headers).httpClient.headers, is(equalTo(combinedHeaders)));
+        assertThat(clientWithNoHeaders2.withHeaders(extraHeaders).httpClient.headers, is(equalTo(extraHeaders)));
+        assertThat(clientWithNoHeaders2.withHeaders(extraHeaders).withHeaders(headers).httpClient.headers, is(equalTo(combinedHeaders)));
+        assertThat(clientWithHeaders.withHeaders(extraHeaders).httpClient.headers, is(equalTo(combinedHeaders)));
+
+        assertThat(clientWithNoHeaders1.httpClient.headers, is(noHeaders));
+        assertThat(clientWithNoHeaders2.httpClient.headers, is(noHeaders));
+        assertThat(clientWithHeaders.httpClient.headers, is(headers));
```
Comment (xerial) [inline src/test/java/com/treasuredata/client/TestTDClientConfig.java]: 👍  for exhaustive test

**Comment by xerial** [comment]
LGTM!  I guess pinging me in slack#mpp is the best way to get my attention :(

File: src/main/java/com/treasuredata/client/AbstractTDClientBuilder.java (hunk 1 above)
Comment (danielnorberg) [inline src/main/java/com/treasuredata/client/AbstractTDClientBuilder.java]: That is a good question. The reason we need this is to increase the request buffer size to accommodate passing on some td console session headers from workflow-api to td-api. And they are big enough to cause http requests to fail with a "too large request" error when using default buffer sizes.

**Pull request merged**